File: C#/Binary Tree - DFS/437. Path Sum III.cs
/* 437. Path Sum III 
Difficulty: Easy
Submission Link: https://leetcode.com/problems/path-sum-iii/submissions/1388938628/ */

public class Solution {
    public int PathSum(TreeNode root, int targetSum) 
    {
        if(root == null) return 0;
        
        return Dfs(root, targetSum) 
            + PathSum(root.left, targetSum)
            + PathSum(root.right, targetSum);

        int Dfs(TreeNode node, long sum)
        {
            var s = node.val == sum ? 1 : 0;

            if(node.left != null)
                s += Dfs(node.left, sum-node.val);

            if(node.right != null)
                s += Dfs(node.right, sum-node.val);

            return s;
        }
    }
}

File: Python3/Binary Tree - DFS/437. Path Sum III.py
""" 437. Path Sum III 
Difficulty: Easy
Submission Link: https://leetcode.com/problems/path-sum-iii/submissions/1388937462/ """

class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        def dfs(node, currSum):
            nonlocal ans, prefixSum
            if node is None:
                return

            currSum += node.val
            ans += prefixSum[currSum - targetSum]
            prefixSum[currSum] += 1
            
            dfs(node.left, currSum)
            dfs(node.right, currSum)
            
            prefixSum[currSum] -= 1

        prefixSum = defaultdict(int)
        prefixSum[0] = 1
        ans = 0
        dfs(root, 0)
        return ans